src/realm/obj.cpp

```diff
@@ -2262,14 +2262,31 @@ ref_type Obj::Internal::get_ref(const Obj& obj, ColKey col_key)
     return to_ref(obj._get<int64_t>(col_key.get_index()));
 }
 
-ref_type Obj::get_collection_ref(Index index) const noexcept
+ref_type Obj::get_collection_ref(Index index, CollectionType type) const
 {
-    return to_ref(_get<int64_t>(mpark::get<ColKey>(index).get_index()));
+    ColKey col_key = mpark::get<ColKey>(index);
+    if (col_key.is_collection()) {
+        return to_ref(_get<int64_t>(col_key.get_index()));
+    }
+    if (col_key.get_type() == col_type_Mixed) {
+        auto val = _get<Mixed>(col_key.get_index());
+        if (val.is_null() || !val.is_type(DataType(int(type)))) {
```

Comment:
I think this check can be moved inside the constructor for each collection? Can't it? In order to avoid having to check if the `col_key` has the right type and also inspect the `Mixed`.

Reply:
`get_collection_ref` is called whenever the collection need to update itself - not during construction. If a DIctionary has been replaced by a list (or vice versa), this will be caught here.